### ISSUE TITLE
bug: parallel read requests are very slow

### ISSUE THREAD

**Issue by apoorvsadana** [keep-starknet-strange/madara#1414]
Currently the RPC layer of Madara can't handle read requests in parallel. As a result, sending multiple read calls on a Madara node takes a lot of time. The issue hasn't been diagnosed yet, possible reasons include a lock which is being taken by read calls or something mentioned [here](https://github.com/paritytech/substrate/issues/8846).

### Expected behaviour
Multiple read calls can be processed in parallel.

**Comment by apoorvsadana** [comment]
It is because 

1. Even though users are reading via full nodes, there will be multiple full nodes syncing on Madara. At the current speed, this will make syncing slow for the full nodes and will be a bottleneck when autoscaling full nodes.
2. For now, till we don't have the p2p spec, the full nodes being used will be Madara nodes. That's what we are planning for the Sharingan testnet.

**Comment by tdelabro** [comment]
>  there will be multiple full nodes syncing on Madara

They are not syncing through RPC calls, are they?

> the full nodes being used will be Madara nodes

More likely they will be deoxys nodes? With some optimization for reading and not for writing

**Comment by apoorvsadana** [comment]
> They are not syncing through RPC calls, are they?
I mean until we have p2p it has to be via the RPC right?

> More likely they will be deoxys nodes? With some optimization for reading and not for writing
Ideally it should be deoxys but that would also need p2p first right?

**Comment by apoorvsadana** [comment]
Also, since deoxys is also a Madara fork I guess it would fall into the same issue? Do you know how this works on current Substrate chains? Because over there it's the same node running as a full node instead of a validator. So we should be missing something otherwise parallel read calls should work by default.

**Comment by tdelabro** [comment]
> I mean until we have p2p it has to be via the RPC right?

We already have p2p. Substrate p2p to share block. I think we will only be missing the sharing of sierra classes, which I will impl soon

**Comment by tdelabro** [comment]
I think deoxys should try to minimize runtime call to only the simulate/estimate/trace ones. And have a big disk storage + cache to serve requests as fast as possible.
In madara we purposefully refuse to store some stuffs and rely to runtime call to do some read. Maybe that can be changed for us too, as long as it doesn't slow down block production.
Overall I agree, the less runtime call the better

**Comment by apoorvsadana** [comment]
Ok update on this. There's not issue as such with the JSON RPC settings. The most time consuming part of our read queries is the `compute_hash` function. I tried this with the pending receipt RPC call. If we compute the txn_hash during the RPC call, it takes somehwere around 130-250ms. On the other hand, if I only remove hash calculation and do something dummy, getting the receipt takes like 10-20ms. I got similar results for getEvents. So we need to optimize our read calls and we can use caching on the client side to cache the txn hashes to make the calls faster.

**Comment by tdelabro** [comment]
We already have some cold storage caching for tx hashes:
```
    /// When enabled, more information about the blocks and their transaction is cached and stored
    /// in the database.
    ///
    /// This may improve response times for RPCs that require that information, but it also
    /// increases the memory footprint of the node.
    #[clap(long)]
    pub cache: bool,
```
crates/node/src/commands/run.rs

Try run the code with `cache=true` and tell us how much improvement comes with it

**Comment by apoorvsadana** [comment]
Yep, I tried. But `getEvents` and `getTransactionReceipt` aren't configured to use this cache so didn't get any improvement with this. Also the current cache is `block_hash => txn_hashes`. For the cache to be used, I am thinking to create a new one, `(block_hash, index) => txn_hash`, in this way we can use them for the getEvents and getTxnReceipt. wdyt?

**Comment by tdelabro** [comment]
It's another storage for the same thing... Not super good.
You can use the existing one and index at your index. It will be fast. And will avoid duplication.

Also, I think we should create an actual cache this time. An in-memory one. We never did it https://github.com/keep-starknet-strange/madara/issues/947

**Comment by apoorvsadana** [comment]
> You can use the existing one and index at your index. It will be fast. And will avoid duplication.
 
 I didn't understand this, do you mean having a mapping like `block_hash => txn_index => txn_hashes` in one column itself. Maybe by converting the Vec<> into a HashMap?
 
 > Also, I think we should create an actual cache this time. An in-memory one. We never did it https://github.com/keep-starknet-strange/madara/issues/947
 
 Oh yes, we should implement. Can we do this in a separate issue though? I think this will take more time and the current cache one can be an easy fix to improve performance by a lot.

**Comment by apoorvsadana** [comment]
@antiyro do you think you can pick up the caching issue in that case when you're done with spec compatibility?

**Comment by tdelabro** [comment]
What I meant is that we already have a `block_hash => txn_hashes` mapping. We don't want to have both this and `(block_hash, index) => txn_hash`. And for this I think it's better to do one big read then iterate in memory, rather than have the one using `block_hash => txn_hashes` do one read per each tx.

So I would say keep `block_hash => txn_hashes` and use it in `getEvents` and `getTransactionReceipt` and any other that need it.

**Comment by tdelabro** [comment]
> Oh yes, we should implement this. Can we do this in a separate issue though? I think this will take more time and the current cache one can be an easy fix to improve performance by a lot.

Sure. But we can keep it in mind

**Comment by antiyro** [comment]
Yes, we'll go through a lot of RPC improvements in the near future, with someone working full time on it.

**Comment by apoorvsadana** [comment]
Closing this issue a we already have #947 and #1427 fixed the current issue for Gomu gomu integration.
